enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+that is in the same namespace as the IngressController and contains a CA
+certificate bundle.  If `ClientCA` is omitted, then the IngressController uses
+the system trust bundle.  If `ClientCA` is specified, then the IngressController
+uses the provided bundle to verify client certificates, and the
+IngressController does not start serving unless and until it can read the CA
+certificate bundle as well as the certificate revocation list (CRL) for each
+certificate therein that specifies one or more destination points for retrieving
+that certificate's CRL.
+
+Finally, the optional `AllowedSubjectPatterns` field may be used to specify a
+list of patterns.  If the field is specified, then the IngressController rejects
+any client certificate that does not match at least one of the provided
+patterns.
+
+The following IngressController enables mandatory use of client certificates,
+verified using the CA certificates in the `router-ca-certs-default` configmap,
+and requiring that client certificates match the provided subject:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  clientTLS:
+    clientCertificatePolicy: Required
+    clientCA:
+      name: router-ca-certs-default
+    allowedSubjectPatterns:
+    - "^/CN=example.com/ST=NC/C=US/O=Security/OU=OpenShift$"
+```
+
+#### Validation
+
+Omitting `spec.clientTLS` has well defined semantics.
+
+The API validates the `spec.clientTLS.clientCertificatePolicy` field value as
+described by the field type's `+kubebuilder:validation:Enum` marker.
+
+The API does not validate that the `spec.clientTLS.clientCA.name` field value
+references a valid configmap; if the named configmap does not exist, then the
+IngressController deployment is not scheduled until the configmap is created.
+
+The API does not validate the `spec.clientTLS.allowedSubjectPatterns` field
+value.
+
+### User Stories
+
+#### As a cluster administrator, I do not want to enable use of client certificates.
+
+Client TLS is not enabled by default, so no action is needed.
+
+#### As a cluster administrator, I want to enable optional use of client certificates.
+
+The user enables client TLS with the `Optional` client certificate policy.  The
+operator configures the IngressController to request and verify client
+certificates using the system trust bundle.
+
+#### As a cluster administrator, I want to require client certificates using a CA that I control.
+
+The user enables client TLS with the `Required` client certificate policy and a
+custom CA certificate bundle.  The operator configures the IngressController to
+request and verify client certificates using the provided CA certificate bundle
+and reject clients that do not present valid client certificates.
+
+#### As a cluster administrator, I want to require client certificates using a CA that I control, respecting its certificate revocation list.
```

Comment:
does HA proxy do this automatically?  What happens if the CRL cannot be retrieved?

Reply:
HAProxy doesn't download the CRL; the ingress operator will download it, publish it to a configmap, and mount the configmap into the pod that has the haproxy process.  If the CRL cannot be retrieved, then the router pods will not start.  If router pods are already running, and the CRL cannot be retrieved, then the pods will continue running with the previously downloaded CRL.